File: cli/go/awid/agents.go
package awid

import "context"

// IdentityView is returned by GET /v1/agents.
type IdentityView struct {
	IdentityID          string `json:"identity_id,omitempty"`
	AgentID             string `json:"agent_id,omitempty"`
	Alias               string `json:"alias"`
	Name                string `json:"name,omitempty"`
	HumanName           string `json:"human_name,omitempty"`
	AgentType           string `json:"agent_type,omitempty"`
	Status              string `json:"status,omitempty"`
	LastSeen            string `json:"last_seen,omitempty"`
	Online              bool   `json:"online"`
	AccessMode          string `json:"access_mode,omitempty"`
	AddressReachability string `json:"address_reachability,omitempty"`
	NamespaceSlug       string `json:"namespace_slug,omitempty"`
	Lifetime            string `json:"lifetime,omitempty"`
}

func (v IdentityView) CurrentIdentityID() string {
	if v.IdentityID != "" {
		return v.IdentityID
	}
	return v.AgentID
}

type ListIdentitiesResponse struct {
	ProjectID  string         `json:"project_id"`
	Identities []IdentityView `json:"identities,omitempty"`
	Agents     []IdentityView `json:"agents,omitempty"`
}

func (c *Client) ListIdentities(ctx context.Context) (*ListIdentitiesResponse, error) {
	var out ListIdentitiesResponse
	if err := c.Get(ctx, "/v1/agents", &out); err != nil {
		return nil, err
	}
	if len(out.Identities) == 0 && len(out.Agents) != 0 {
		out.Identities = out.Agents
	}
	return &out, nil
}

type PatchIdentityRequest struct {
	AccessMode          string `json:"access_mode,omitempty"`
	AddressReachability string `json:"address_reachability,omitempty"`
}

type PatchIdentityResponse struct {
	IdentityID          string `json:"identity_id,omitempty"`
	AgentID             string `json:"agent_id,omitempty"`
	AccessMode          string `json:"access_mode,omitempty"`
	AddressReachability string `json:"address_reachability,omitempty"`
}

func (r PatchIdentityResponse) CurrentIdentityID() string {
	if r.IdentityID != "" {
		return r.IdentityID
	}
	return r.AgentID
}

func (c *Client) PatchIdentity(ctx context.Context, identityID string, req *PatchIdentityRequest) (*PatchIdentityResponse, error) {
	var out PatchIdentityResponse
	if err := c.Patch(ctx, "/v1/agents/"+urlPathEscape(identityID), req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// HeartbeatResponse is returned by POST /v1/agents/heartbeat.
type HeartbeatResponse struct {
	IdentityID string `json:"identity_id"`
	LastSeen   string `json:"last_seen"`
	TTL        int    `json:"ttl_seconds"`
}

// Heartbeat reports agent liveness to the aweb server.
func (c *Client) Heartbeat(ctx context.Context) (*HeartbeatResponse, error) {
	var out HeartbeatResponse
	if err := c.Post(ctx, "/v1/agents/heartbeat", nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

File: server/src/aweb/mcp/tools/agents.py
"""MCP tools for agent listing and heartbeat."""

from __future__ import annotations

import json
from uuid import UUID

from aweb.mcp.auth import get_auth
from aweb.presence import (
    DEFAULT_PRESENCE_TTL_SECONDS,
    list_agent_presences_by_ids,
    update_agent_presence,
)


async def list_agents(db_infra, redis) -> str:
    """List all agents in the authenticated project."""
    auth = get_auth()
    aweb_db = db_infra.get_manager("aweb")

    rows = await aweb_db.fetch_all(
        """
        SELECT agent_id, alias, human_name, agent_type, access_mode,
               did, custody, lifetime, status
        FROM {{tables.agents}}
        WHERE project_id = $1 AND deleted_at IS NULL AND agent_type != 'human'
        ORDER BY alias
        """,
        UUID(auth.project_id),
    )

    agent_ids = [str(r["agent_id"]) for r in rows]
    presences = await list_agent_presences_by_ids(redis, agent_ids)
    presence_map = {p["agent_id"]: p for p in presences}

    agents = []
    for r in rows:
        aid = str(r["agent_id"])
        p = presence_map.get(aid)
        agents.append(
            {
                "agent_id": aid,
                "alias": r["alias"],
                "human_name": r.get("human_name") or "",
                "agent_type": r.get("agent_type") or "agent",
                "access_mode": r.get("access_mode", "open"),
                "online": p is not None,
                "did": r.get("did") or "",
                "custody": r.get("custody") or "",
                "lifetime": r.get("lifetime") or "ephemeral",
                "status": r.get("status") or "active",
            }
        )

    return json.dumps({"project_id": auth.project_id, "agents": agents})


async def heartbeat(db_infra, redis) -> str:
    """Send a heartbeat to maintain agent presence."""
    auth = get_auth()
    aweb_db = db_infra.get_manager("aweb")

    row = await aweb_db.fetch_one(
        """
        SELECT alias
        FROM {{tables.agents}}
        WHERE agent_id = $1 AND project_id = $2 AND deleted_at IS NULL
        """,
        UUID(auth.agent_id),
        UUID(auth.project_id),
    )
    if not row:
        return json.dumps({"error": "Agent not found"})

    ttl = DEFAULT_PRESENCE_TTL_SECONDS
    last_seen = await update_agent_presence(
        redis,
        agent_id=auth.agent_id,
        alias=row["alias"],
        project_id=auth.project_id,
        ttl_seconds=ttl,
    )

    return json.dumps(
        {
            "agent_id": auth.agent_id,
            "last_seen": last_seen,
            "ttl_seconds": ttl,
        }
    )
